config.c: #tryinclude should not create new files

Allan-N opened the pull request:
Correct an issue in ast_config_text_file_save2() when updating configuration files with "#tryinclude" statements. The API currently replaces "#tryinclude" with "#include". The API also creates empty template files if the referenced files do not exist. This change resolves these problems.

Allan-N commented:
This change resolves Issue #920

Allan-N commented:
Note : I was asked about the "precomment" changes included in this PR.  To set the stage, let's talk about a sample configuration file :
```
 1. ; comment block #1
 2. ; more of comment block #1
 3. 
 4. [category]
 5. 
 6. ; comment block #2
 7. ; more of comment block #2
 8. 
 9. variable1 = value1	; inline comment
10.			; more of inline comment
11.
12. ; comment block #3
13. ; more of comment block #3
14. 
15. #include somefile
16. 
17. ; comment block #4
18. ; more of comment block #4
19. 
20. variable2 = value2
```
The pre-PR code would have lines 1-3 associated with "[category]" as pre-comments, lines 5-8 as pre-comments to "variable1", and lines 11-14 AND lines 16-19 as pre-comments to "variable2".  As-is, the code would rewrite the above as :
```
 1. ; comment block #1
 2. ; more of comment block #1
 3. 
 4. [category]
 5. 
 6. ; comment block #2
 7. ; more of comment block #2
 8. 
 9. variable1 = value1	; inline comment
10.			; more of inline comment
11. #include somefile
12.
13. ; comment block #3
14. ; more of comment block #3
15. 
16. 
17. ; comment block #4
18. ; more of comment block #4
19. 
20. variable2 = value2
```
The "precomment" changes included in this PR are to keep the comments prior to the "#include" in place.  I think this fix is essential.  The bonus change I included (lines 2559-2570) relates to the handling of leading whitespace on standalone and inline comments.

InterLinked1 reviewed (changes requested):
All your changes should be squashed into a single commit (amend your original commit when making edits, rather than making a new commit)

Allan-N commented:
>All your changes should be squashed into a single commit

Done.

seanbright commented on main/config.c:
```diff
@@ -255,7 +255,16 @@ struct ast_config {
+	struct ast_config_include *includes_last;
+};
+
+struct ast_config_included_file {
```
Avoid the `ast_` prefix when creating new things that will not be exported.

Allan-N replied:
Poof.  The `ast_` prefix has been removed

seanbright commented on main/config.c:
```diff
@@ -265,7 +274,7 @@ struct ast_config_include {
 	char *include_location_file;
 	int  include_location_lineno;    /*!< lineno where include occurred */
-	int  exec;                       /*!< set to non-zero if its a #exec statement */
+	int  include_type;               /*!< set positive if #exec statement, negative if #tryinclude statement */
```
This structure is opaque, so this type can be changed to an `enum` to be more expressive.

Allan-N replied:
Not knowing for certain if there was anyone calling ast_include_new() with a value other than 0 or 1 made me a bit hesitant about using an enum.  But, I agree that using an enum makes the code more expressive.

Enum added, code updated.

Allan-N commented:
>The commit message should wrap at 72 characters.

Done.

> Also, the title could maybe be a bit more descriptive, like "config.c: Fix #tryinclude being converted to #include on rewrite" or something.

Updated.

> Finally, the commit message should contain a "Resolves" line closing the linked issue (number 920 in this case).

Added.

InterLinked1 commented:
> > The commit message should wrap at 72 characters.
> 
> Done.
> 
> > Also, the title could maybe be a bit more descriptive, like "config.c: Fix #tryinclude being converted to #include on rewrite" or something.
> 
> Updated.
> 
> > Finally, the commit message should contain a "Resolves" line closing the linked issue (number 920 in this case).
> 
> Added.

Looks like you updated the PR description, not the commit itself (which is what matters).

Allan-N commented:
>Looks like you updated the PR description, not the commit itself (which is what matters).

I have now updated the commit too :-)

InterLinked1 commented:
> > Looks like you updated the PR description, not the commit itself (which is what matters).
> 
> I have now updated the commit too :-)

`This change resolves Issue #920` should be `Resolves: #920` as GitHub looks for specific syntax to autoclose the issue

seanbright commented on include/asterisk/config.h:
```diff
@@ -59,6 +59,18 @@ enum config_save_flags {
 #define	CONFIG_STATUS_FILEUNCHANGED	(void *)-1
 #define	CONFIG_STATUS_FILEINVALID	(void *)-2
 
+/*!
+ * \brief Types used for ast_include_new() include_type
+ */
+typedef enum {
+	/*! #include statements */
+	CONFIG_STATEMENT_INCLUDE    = 0,
+	/*! #exec statements */
+	CONFIG_STATEMENT_EXEC       = 1,
+	/*! #tryinclude statements */
+	CONFIG_STATEMENT_TRYINCLUDE = 2,
+} include_statement_type;
```
I think my previous comments were confusing.

You should add this `enum` only in `main/config.c` - `ast_include_new()` will still be forced to take an `int` for API compatibility.

Allan-N replied:
Understood.  I changed the `ast_include_new()` argument type back to an `int` and moved the `enum` to `main/config.c`.

Allan-N commented:
cherry-pick-to: 18
cherry-pick-to: 20
cherry-pick-to: 21
cherry-pick-to: 22

InterLinked1 commented on main/config.c:
```diff
@@ -362,20 +390,65 @@ struct ast_config_include *ast_include_new(struct ast_config *conf, const char *
+		}
+		break;
+	case CONFIG_STATEMENT_TRYINCLUDE :
+	    {
```
whitespace

Allan-N replied:
This is a 1/2 indent that follows the coding style already present in the file (look for "case PARSE_INADDR")

InterLinked1 commented on main/config.c:
```diff
@@ -362,20 +390,65 @@ struct ast_config_include *ast_include_new(struct ast_config *conf, const char *
+			AST_LIST_INSERT_TAIL(&inc->included_files, inc_file, next);
+		}
+		break;
+	case CONFIG_STATEMENT_TRYINCLUDE :
```
shouldn't be a space before the colons here and elsewhere

Allan-N replied:
Fixed

InterLinked1 commented on main/config.c:
```diff
@@ -255,17 +255,38 @@ struct ast_config {
+/*!
+ * \brief Types used for ast_include_new() include_type
+ */
+typedef enum {
```
Do not make a typedef for this, just use this as an enum directly: https://docs.asterisk.org/Development/Policies-and-Procedures/Coding-Guidelines/#dont-use-unnecessary-typedefs

Allan-N replied:
Fixed

InterLinked1 reviewed (changes requested):
> And, as a bonus, I also found/fixed the issue where standalone and inline comments in the .conf files were being stripped of all leading whitespace.

Personally, I think this should be broken out into a separate patch/PR, since it's not related to the primary purpose of the patch.

Allan-N commented:
>>And, as a bonus, I also found/fixed the issue where standalone and inline comments in the .conf files were being stripped of all leading whitespace.

>Personally, I think this should be broken out into a separate patch/PR, since it's not related to the primary purpose of the patch.

I removed my "bonus" fix and will create a new issue/PR to address the issue(s) with configuration file rewrites mangling most of the formatted comments.

asteriskteam commented:
This PR has been marked stale because it has been in "Changes Requested" or "submitter-action-required" state for 28 days or more. Please make the requested changes within 14 days or the PR will be closed.

Allan-N commented:
@seanbright, @InterLinked1 : I believe that all requested changes have been made.  Can you kindly review?

asteriskteam commented:
This PR has been marked stale because it has been in "Changes Requested" or "submitter-action-required" state for 28 days or more. Please make the requested changes within 14 days or the PR will be closed.

Allan-N commented:
[PR #1047](https://github.com/asterisk/asterisk/pull/1047) pulled in one of the fixes (a reference leak) that I had found/addressed with my change.  I have since rebased and updated this pull request.

I believe that all of the previously requested changes have been made. Can you kindly review / approve / ... ?

InterLinked1 commented on include/asterisk/config.h:
```diff
@@ -914,7 +914,7 @@ void ast_category_rename(struct ast_category *cat, const char *name);
 #define ast_variable_new(name, value, filename) _ast_variable_new(name, value, filename, __FILE__, __PRETTY_FUNCTION__, __LINE__)
 
-struct ast_config_include *ast_include_new(struct ast_config *conf, const char *from_file, const char *included_file, int is_exec, const char *exec_file, int from_lineno, char *real_included_file_name, int real_included_file_name_size);
+struct ast_config_include *ast_include_new(struct ast_config *conf, const char *from_file, const char *included_file, int include_type, const char *exec_file, int from_lineno, char *real_included_file_name, int real_included_file_name_size);
```
I think it would be nice for argument `include_type` to have the more precise type `enum include_statement_type`, not `int`.

I know this is modifying a public API, but I think it should still work out to be backwards-compatible since you assigned the same values to the enum.

InterLinked1 replied:
Actually, now that I think about it, I'm not sure how the compiler would like passing ints if the parameter type is enum... if it complains about that, then maybe ignore this, or make the change master only.

seanbright replied:
I _think_ going from `int` to `enum` would be fine ABI-wise (I believe it is possible that the compiler could choose a smaller type), but best not to change it in release branches.

If we're going to change the parameter type, best to do do it in a separate PR and maybe add some doxygen while we're at it.

Allan-N replied:
When I first started working on these changes I was concerned about changing the parameter type.  As part of the review process with @seanbright (on 10/03) I changed the parameter type to an `enum` (defined in `include/asterisk/config.h` and used in `main/config.c`).  A followup review comment (on 10/13) suggested that the `enum` should be only in `main/config.c` ... so I made that change.

Let me know if you want me to switch back to using an `enum` in both the .h and .c files now, create a new PR making the change later (after this change has been merged), or leave this PR as-is.

InterLinked1 commented on main/config.c:
```diff
@@ -3019,12 +3163,9 @@ int ast_config_text_file_save2(const char *configfile, const struct ast_config *
-			fclose(f);
+			print_include(f, incl);
 			incl->output = 1;
+			fclose(f);
```
Why switch the order here?

Allan-N replied:
For consistency.  Every other place were I call `print_include(f, incl);` is followed with `incl->output = 1;`

seanbright commented:
> I believe that all of the previously requested changes have been made. Can you kindly review / approve / ... ?

Yup, that is what I was doing when I saw the reference count leak you identified. Will provide feedback shortly.

seanbright commented on main/config.c:
```diff
@@ -2750,15 +2868,30 @@ int ast_config_text_file_save2(const char *configfile, const struct ast_config *
-			if (!is_writable(fn)) {
+		if (incl->include_type != CONFIG_STATEMENT_EXEC) {
+			struct config_included_file *included_file;
+			int ok = 0;
```
`int ok = 1;`

seanbright commented on main/config.c:
```diff
@@ -2750,15 +2868,30 @@ int ast_config_text_file_save2(const char *configfile, const struct ast_config *
+			AST_LIST_TRAVERSE(&incl->included_files, included_file, next) {
+				if (!is_writable(included_file->path)) {
+					/* the #tryinclude path exists but is not writable */
+					ok = -1;
```
`ok = 0;`

seanbright commented on main/config.c:
```diff
@@ -2750,15 +2868,30 @@ int ast_config_text_file_save2(const char *configfile, const struct ast_config *
+				}
+			}
+
+			if (ok != 0) {
```
`if (!ok) {`

seanbright reviewed (changes requested):
Here's a handful of things I've seen so far. Still working through it/testing/etc.

seanbright reviewed (changes requested):
So before your patch, this file:

```
[general]
a = b
c = d

; I'm just a comment

; #include "issue920/inc.conf"
; #include "issue920/glob/inc-*.conf"
#tryinclude "issue920/tryinc.conf"
#tryinclude "issue920/glob/tryinc-*.conf"
```

Saves as (ignoring the 'generated' header):

```
[general]
a = b
c = d
; I'm just a comment

; #include "issue920/inc.conf"
; #include "issue920/glob/inc-*.conf"
#include "issue920/glob/tryinc-*.conf"
#include "issue920/tryinc.conf"
```

After your patch, it saves as:

```
[general]
a = b
c = d
#tryinclude "issue920/tryinc.conf"
#tryinclude "issue920/glob/tryinc-*.conf"
```

So the `#tryinclude`s render correctly, but we lose the comments.

Allan-N commented:
> Saves as (ignoring the 'generated' header):
> ...
> After your patch, it saves as:
> ...

Those results do not match what I've been seeing with my testing.  But, regardless, I will make the requested updates and re-check the behavior with the handling of comments as you should be able to write the config with the file unchanged, modulo the updated header.

Stay tuned.

Allan-N commented:
> So the #tryincludes render correctly, but we lose the comments.

Well, it turns out that the .conf I was using for testing worked but your test case presented a different (and very reasonable) scenario ... and one that I had not accounted for in my changes.  That has since been corrected!  This PR now has a bit more but with some added refactoring the changes are a bit cleaner.

seanbright commented on main/config.c:
```diff
@@ -363,20 +392,65 @@ struct ast_config_include *ast_include_new(struct ast_config *conf, const char *
 	if (!inc->include_location_file
 		|| !inc->included_file
-		|| (is_exec && !inc->exec_file)) {
+		|| ((include_type == CONFIG_STATEMENT_EXEC) && !inc->exec_file)) {
```
You moved the `inc->exec_file = ast_strdup(exec_file);` so that it occurs after this check. I haven't tested but I assume this means we won't get the `#exec` rendered in to the saved file.

seanbright commented on main/config.c:
```diff
@@ -363,20 +392,65 @@ struct ast_config_include *ast_include_new(struct ast_config *conf, const char *
+		break;
+	    }
+	case CONFIG_STATEMENT_EXEC:
+		inc->exec_file = ast_strdup(exec_file);
```
I think anywhere in this function where memory allocation fails we should immediately return `NULL` (calling the necessary cleanup function first).

I realize there is existing code in this function that does not check for allocation failure, but there is no reason to keeping kicking that can down the road.

Allan-N replied:
Fixed

seanbright commented on main/config.c:
```diff
@@ -363,20 +392,65 @@ struct ast_config_include *ast_include_new(struct ast_config *conf, const char *
 	/* attach this new struct to the conf struct */
-	inc->next = conf->includes;
-	conf->includes = inc;
+	if (conf->includes_root && conf->includes_last) {
+		conf->includes_last->next = inc;
+	} else {
+		conf->includes_root = inc;
+	}
+	conf->includes_last = inc;
```
Would prepending the list change semantics? That would avoid the need for the extra pointer. (minor)

Allan-N replied:
Yes, we follow the list when the actual #include lines are written to the .conf file.

seanbright reviewed (changes requested):
FWIW - This is a difficult PR to review because it both fixes the `#include` / `#tryinclude` issue as well as the comment rendering. I'm not suggesting that you split it up, just justifying the number of rounds of review involved.

Allan-N commented:
>FWIW - This is a difficult PR to review because it both fixes the #include / #tryinclude issue as well as the comment rendering. I'm not suggesting that you split it up, just justifying the number of rounds of review involved.

Yes, I realize that this is not a simple change and I very much appreciate your willingness to review the change and fixing an issue.

The gotcha is that the comments had been associated with the "category" or "variable".  There was no way to track a #include nested within pre-category or pre-variable comments.  Now, we track the comments before and after the #include's.  Now, you can also have comments before the #include (describing what's being included) and comments after the #include (describing whatever category/variable/end-of-file that follows),

seanbright commented on main/config.c:
```diff
@@ -363,20 +392,70 @@ struct ast_config_include *ast_include_new(struct ast_config *conf, const char *
-		inc->exec_file = ast_strdup(exec_file);
-
 	if (!inc->include_location_file
 		|| !inc->included_file
-		|| (is_exec && !inc->exec_file)) {
+		|| ((include_type == CONFIG_STATEMENT_EXEC) && !exec_file)) {
 		ast_includes_destroy(inc);
 		return NULL;
 	}
```
These checks were originally checking for allocation failures for the previous 3 potential calls to `ast_strdup`, so checking `exec_file` here doesn't make sense. Remove the entire exec expression from here and check for allocation failure in the `CONFIG_STATEMENT_EXEC` branch of the switch statement below instead.

Allan-N replied:
You are correct.  I have updated the allocation failure checks here and in the `CONFIG_STATEMENT_EXEC` branch.

seanbright commented on main/config.c:
```diff
@@ -2125,17 +2232,31 @@ static int process_text_line(struct ast_config *cfg, struct ast_category **cat,
 
-		do_include = ast_config_internal_load(cur, cfg, flags, real_inclusion_name, who_asked) ? 1 : 0;
+		result = ast_config_internal_load(cur, cfg, flags, real_inclusion_name, who_asked);
 		if (!ast_strlen_zero(exec_file))
 			unlink(exec_file);
```
This used to happen regardless of the result of `ast_include_new` but now if `ast_include_new` fails we return early and this doesn't happen.

Allan-N replied:
Fixed!

seanbright commented on main/config.c:
```diff
@@ -2737,6 +2864,41 @@ static int is_writable(const char *fn)
+{
+	switch (incl->include_type) {
+	case CONFIG_STATEMENT_INCLUDE:
+		fprintf(f,"#include \"%s\"\n", incl->included_file);
```
Make one more pass through your changes and sprinkle in some spaces where appropriate for coding guidelines. So here after `f,`. I'll highlight some more but not all of them. Mostly I've seen them here and in `for` loops.

Allan-N replied:
I made a pass through the entire file.  Most certainly added the space after a "," in my changes ... and likely a few extra places too :-)

seanbright commented on main/config.c:
```diff
@@ -2832,31 +3014,24 @@ int ast_config_text_file_save2(const char *configfile, const struct ast_config *
 			/* dump any includes that happen before this category header */
-			for (incl=cfg->includes; incl; incl = incl->next) {
-				if (strcmp(incl->include_location_file, cat->file) == 0){
+			for (incl=cfg->includes_root; incl; incl = incl->next) {
```
Spaces here around the `=`

seanbright commented on main/config.c:
```diff
@@ -2937,24 +3113,23 @@ int ast_config_text_file_save2(const char *configfile, const struct ast_config *
-				for (incl=cfg->includes; incl; incl = incl->next) {
-					if (strcmp(incl->include_location_file, var->file) == 0){
+				/* dump any includes that happen before this variable */
+				for (incl=cfg->includes_root; incl; incl = incl->next) {
```
Spaces here around the `=`

seanbright commented on main/config.c:
```diff
@@ -2990,6 +3163,42 @@ int ast_config_text_file_save2(const char *configfile, const struct ast_config *
+
+		/* Now, for files with trailing #include/#exec statements,
+		   we have to make sure every entry is output */
+		for (incl=cfg->includes_root; incl; incl = incl->next) {
```
Spaces here around the `=`

seanbright reviewed (changes requested):
A couple more small things and then I'm happy.
